### PR TITLE
🎉 Source FB Marketing: add `time_increment` parameter to custom insight streams

### LINKED ISSUE #6225: Source Facebook Marketing:  Allow user to set `time_increment` parameter for Insights API
## Tell us about the problem you're trying to solve
In the current behavior of this connector, the call to the Insights API returns data in single day increments.  

However, the field `reach` doesn't sum correctly over days, so there is no way to correctly determine the reach of your ads over a greater time period.  The problem also affects the parameter `frequency`.   In order to have the correct reach over a greater time period, you would need to be able to ask for metrics aggregated over a greater "time slice".  

## Describe the solution you’d like
The Insights API supports a request for a broader aggregate of metrics over an extended time slice through the parameter `time_increment`, 

https://developers.facebook.com/docs/marketing-api/insights/parameters/v12.0

There is currently no option for the user to set this parameter.  Being able to set it to `monthly, all_days` or an integer for the source would solve the problem.  

## Describe the alternative you’ve considered or used
I have not found a workaround to correctly determine ad reach over a period of time.

## Are you willing to submit a PR?

Not immediately

### PR DESCRIPTION
**Pull request by keu**
## What

closing #6225 

## How
When we slice date intervals it is important to stick to the start point - `start_date`. 
This is why we define three new parameters on the custom insight level:
- `start_date` with default to global `start_date`
- `end_date` with default to global `end_date`
- `time_increment` with default `1`, and maximum to `90`

## 🚨 User Impact 🚨
No

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

### CODE REVIEW

**Review by sergei-solonitcyn — APPROVED** [review]

**Comment by keu** [comment]
@sherifnada did you have a chance to review this?

File: airbyte-integrations/connectors/source-facebook-marketing/integration_tests/test_streams.py [hunk 1]
```diff
@@ -32,16 +31,22 @@ def state_with_include_deleted_fixture(state):
...
 
 
 class TestFacebookMarketingSource:
     @pytest.mark.parametrize(
         "stream_name, deleted_id", [("ads", "23846756820320398"), ("campaigns", "23846541919710398"), ("ad_sets", "23846541706990398")]
     )
     def test_streams_with_include_deleted(self, stream_name, deleted_id, config_with_include_deleted, configured_catalog):
-        catalog = self.slice_catalog(configured_catalog, {stream_name})
+        catalog = self._slice_catalog(configured_catalog, {stream_name})
+        catalog.streams[0].sync_mode = SyncMode.incremental
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-facebook-marketing/integration_tests/test_streams.py:49]: could you add some context for why we're choosing the 0th stream to be incremental? it's not clear why this is necessary

File: airbyte-integrations/connectors/source-facebook-marketing/integration_tests/test_streams.py [hunk 2]
```diff
@@ -32,16 +31,22 @@ def state_with_include_deleted_fixture(state):
 
 
 @pytest.fixture(scope="session", name="configured_catalog")
-def configured_catalog_fixture():
-    return ConfiguredAirbyteCatalog.parse_file("integration_tests/configured_catalog_without_insights.json")
+def configured_catalog_fixture(config) -> ConfiguredAirbyteCatalog:
+    catalog = SourceFacebookMarketing().discover(logger=logging.getLogger("airbyte"), config=config)
+    streams = [
+        ConfiguredAirbyteStream(stream=stream, sync_mode=stream.supported_sync_modes[0], destination_sync_mode=DestinationSyncMode.append)
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-facebook-marketing/integration_tests/test_streams.py:37]: why are we always using `supported_sync_modes[0]`? is the idea that we use incremental if available otherwise full_refresh? if so why not do something like 

```
# Prefer incremental if available
sync_mode="incremental" if "incremental" in stream.supported_sync_modes else "full_refresh"
```

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/base_insight_streams.py [hunk 3]
```diff
@@ -115,18 +117,24 @@ def state(self) -> MutableMapping[str, Any]:
...
+                "time_increment": self.time_increment,
             }
 
         return {}
 
     @state.setter
     def state(self, value: Mapping[str, Any]):
-        """State setter"""
+        """State setter, will ignore saved state if time_increment is different from previous."""
+        if value.get("time_increment", 1) != self.time_increment:
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/base_insight_streams.py:134]: could you add some context on why we skip? not setting state at all seems like it would contradict the user/developer's expectations no?

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.py [hunk 4]
```diff
@@ -23,11 +23,55 @@
...
+
     action_breakdowns: Optional[List[ValidActionBreakdowns]] = Field(
-        description="A list of chosen action_breakdowns for action_breakdowns", default=[]
+        title="Action Breakdowns",
+        description="A list of chosen action_breakdowns for action_breakdowns",
+        default=[],
+    )
+
+    time_increment: Optional[PositiveInt] = Field(
+        title="Time Increment",
```
[the remark this answers is not in the document]
Comment (keu) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.py:53]: it is not lookback, it grouping period, more I think about it the more I agree `time increment` is better name, but maybe `group_peroiod` ?

File: airbyte-integrations/connectors/source-facebook-marketing/integration_tests/test_streams.py (hunk 1 above)
Comment (keu) [reply airbyte-integrations/connectors/source-facebook-marketing/integration_tests/test_streams.py:49]: this is necessary because it is a list of a single stream, will add comment

File: airbyte-integrations/connectors/source-facebook-marketing/integration_tests/test_streams.py (hunk 2 above)
Comment (keu) [reply airbyte-integrations/connectors/source-facebook-marketing/integration_tests/test_streams.py:37]: no, the idea is to use explicit sync mode, but I see that in some places we just leave it as is

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/base_insight_streams.py (hunk 3 above)
Comment (keu) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/base_insight_streams.py:134]: it is not, as we discussed earlier in other tickets #2154 if user changes connector settings its state is not valid in most of the cases, this is not supported on Airbyte platform level (UX/Core) so we do this at least on connector level.

File: airbyte-integrations/connectors/source-facebook-marketing/integration_tests/test_streams.py (hunk 2 above)
Comment (keu) [reply airbyte-integrations/connectors/source-facebook-marketing/integration_tests/test_streams.py]: updated

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sherifnada) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.py:54]: I currently don't know what this field is referring to. Could you ELI5? (in the description/title)

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.py [hunk 5]
```diff
@@ -10,24 +10,69 @@
...
+        description="The date from which you'd like to replicate data for this stream, in the format YYYY-MM-DDT00:00:00Z.",
+        pattern=DATE_TIME_PATTERN,
+        examples=["2017-01-25T00:00:00Z"],
+    )
+
+    end_date: Optional[datetime] = Field(
+        title="End Date",
+        description=(
+            "The date until which you'd like to replicate data for this stream, in the format YYYY-MM-DDT00:00:00Z. "
+            "All data generated between start_date and this date will be replicated. "
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.py:71]:
```suggestion
            "All data generated between the start date and this date will be replicated. "
```

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/base_insight_streams.py (hunk 3 above)
Comment (sherifnada) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/base_insight_streams.py:134]:
```suggestion
        # if the time increment configured for this stream is different from the one in the previous state
        # then the previous state object is invalid and we should start replicating data from scratch
        # to achieve this, we skip setting the state 
        if value.get("time_increment", 1) != self.time_increment:
```

**Review by sherifnada — APPROVED** [review]
Code LGTM but let' agree on the UX for the spec before merging

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (keu) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.py:54]: @sherifnada please my update, WDYT?

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.py [hunk 6]
```diff
@@ -10,24 +10,72 @@
...
-        description="A list of chosen action_breakdowns for action_breakdowns", default=[]
+        title="Action Breakdowns",
+        description="A list of chosen action_breakdowns for action_breakdowns",
+        default=[],
+    )
+
+    time_increment: Optional[PositiveInt] = Field(
+        title="Time Increment",
+        description=(
+            "Time window in days to get statistic from. The sync will chunk period between the start_date and the end_date by the number of days you specified. "
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.py:56]:
```suggestion
            "Time window in days by which to aggregate statistics. The sync will be chunked into N day intervals, where N is the number of days you specified. "
```

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.py [hunk 7]
```diff
@@ -10,24 +10,72 @@
...
+            "For example if you specify time_increment = 7 then all statistics will be aggregated by 7 days starting from the start_date."
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.py:57]:
```suggestion
            "For example if you set this value to 7, then all statistics will be reported as 7-day aggregates by starting from the start_date. If the start and end dates are October 1st and October 30th, then the connector will output 30 records, each of which represents aggregated stats from a 7 day lookback period."
```

**Review by sherifnada — APPROVED** [review]
@keu added some wording suggestions. Thanks for adding the description!

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.py (hunk 7 above)
Comment (keu) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.py:57]: this is not true, we will have 5 periods:
- 01 - 06
- 07 - 13
- 14 - 20
- 21 - 27
- 28 - 30 ( 3 days)

Comment (sherifnada) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.py:57]: Ah. Can you just edit the comment suggestion to say that instead?
